tests/unit/igraph_circulant.c

```diff
@@ -0,0 +1,195 @@
+/*
+   IGraph library.
+   Copyright (C) 2021  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <igraph.h>
+
+#include "test_utilities.inc"
+
+int main() {
+
+    igraph_t graph, graph_test;
+    igraph_vector_int_t l;
+    igraph_bool_t iso;
+    igraph_integer_t i;
+
+    /* Testing invalid values */
+
+    /* n = -3, l = [1], undirected */
+    igraph_vector_int_init(&l, 0);
+    igraph_vector_int_push_back(&l, 1);
+    CHECK_ERROR(igraph_circulant(&graph, -3, &l, IGRAPH_UNDIRECTED), IGRAPH_EINVAL);
+    igraph_vector_int_destroy(&l);
+
+    /* Testing n = 0 case */
+    /* n = 0, l = [1], undirected */
+
+    igraph_vector_int_init(&l, 0);
+    igraph_vector_int_push_back(&l, 1);
+    IGRAPH_ASSERT(igraph_circulant(&graph, 0, &l, IGRAPH_UNDIRECTED) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_vcount(&graph) == 0);
+    IGRAPH_ASSERT(igraph_ecount(&graph) == 0);
+    igraph_vector_int_destroy(&l);
+
+
+    /* Testing n = 1 case */
+    /* n = 1, l = [1], undirected */
+    igraph_vector_int_init(&l, 0);
+    igraph_vector_int_push_back(&l, 1);
+    IGRAPH_ASSERT(igraph_circulant(&graph, 1, &l, IGRAPH_UNDIRECTED) == IGRAPH_SUCCESS);
```

Comment:
You have allocated a graph on line 43. Here, you re-allocate on top of it, without having destroyed it first. This is why some tests fail:

https://dev.azure.com/igraph-team/igraph/_build/results?buildId=1753&view=logs&j=4c3a8381-b9de-5463-904f-918810769be8&t=a906eab8-d2a3-5e54-d199-606dbbfb0805&l=205

It's good to look at the `linux_static_vendored` test because it has various sanitizers enabled, and it prints stack traces even for non-sanitizer errors.

Reply:
Ah, thanks for finding that for me. I had accidently assumed it was another linux_alpine issue, so didn't look deeper. The graph has been properly destroyed now.